File: packages/nexus-routing/src/utils/logger.ts
/**
 * Unified Logger for Nexus Routing Package
 * Winston-based logging with structured output
 */

import winston from 'winston';

const logLevel = process.env.LOG_LEVEL || 'info';
const mcpMode = process.env.MCP_MODE === 'true';

// Create logger instance
export const logger = winston.createLogger({
  level: logLevel,
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  defaultMeta: { service: 'nexus-routing' },
  transports: [
    // Console transport - use stderr in MCP mode to avoid interfering with JSON-RPC on stdout
    new winston.transports.Console({
      stderrLevels: mcpMode ? ['error', 'warn', 'info', 'debug'] : ['error'],
      format: winston.format.combine(
        winston.format.colorize(),
        winston.format.printf(({ timestamp, level, message, ...meta }) => {
          // In MCP mode, only log errors to stderr
          if (mcpMode && level !== 'error') {
            return '';
          }

          const metaStr = Object.keys(meta).length > 0
            ? ` ${JSON.stringify(meta)}`
            : '';

          return `${timestamp} [${level}]: ${message}${metaStr}`;
        })
      )
    })
  ]
});

// Suppress all logs in strict MCP mode except errors
if (mcpMode && logLevel === 'error') {
  logger.info = () => logger;
  logger.debug = () => logger;
  logger.warn = () => logger;
}

export default logger;

File: services/nexus-fileprocess/worker/internal/logging/logger.go
package logging

import (
	"fmt"
	"log"
	"os"
)

// Logger provides structured logging for the worker
type Logger struct {
	prefix string
	logger *log.Logger
}

// NewLogger creates a new logger with a prefix
func NewLogger(prefix string) *Logger {
	return &Logger{
		prefix: prefix,
		logger: log.New(os.Stdout, fmt.Sprintf("[%s] ", prefix), log.LstdFlags),
	}
}

// Info logs an informational message with key-value pairs
func (l *Logger) Info(msg string, keysAndValues ...interface{}) {
	l.logWithKV("INFO", msg, keysAndValues...)
}

// Warn logs a warning message with key-value pairs
func (l *Logger) Warn(msg string, keysAndValues ...interface{}) {
	l.logWithKV("WARN", msg, keysAndValues...)
}

// Error logs an error message with key-value pairs
func (l *Logger) Error(msg string, keysAndValues ...interface{}) {
	l.logWithKV("ERROR", msg, keysAndValues...)
}

// Debug logs a debug message with key-value pairs
func (l *Logger) Debug(msg string, keysAndValues ...interface{}) {
	l.logWithKV("DEBUG", msg, keysAndValues...)
}

func (l *Logger) logWithKV(level, msg string, keysAndValues ...interface{}) {
	kvStr := ""
	for i := 0; i < len(keysAndValues); i += 2 {
		if i+1 < len(keysAndValues) {
			kvStr += fmt.Sprintf(" %v=%v", keysAndValues[i], keysAndValues[i+1])
		}
	}
	l.logger.Printf("[%s] %s%s", level, msg, kvStr)
}
